src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -134,13 +177,136 @@ private void SetSubFieldNodes(ExecutionContext context, ObjectExecutionNode pare
+        {
+            fields ??= new Fields();
+            List<string> visitedFragmentNames = null;
+            CollectFields(context, specificType.GetNamedType(), selectionSet, fields, ref visitedFragmentNames);
+            return fields;
+
+            void CollectFields(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Fields fields, ref List<string> visitedFragmentNames) //TODO: can be completely eliminated? see Fields.Add
+            {
+                if (selectionSet != null)
+                {
+                    foreach (var selection in selectionSet.SelectionsList)
+                    {
+                        if (selection is Field field)
+                        {
+                            if (ShouldIncludeNode(context, field))
+                                fields.Add(field);
+                        }
+                        else if (selection is FragmentSpread spread)
+                        {
+                            if (visitedFragmentNames?.Contains(spread.Name) != true && ShouldIncludeNode(context, spread))
+                            {
+                                (visitedFragmentNames ??= new List<string>()).Add(spread.Name);
+
+                                var fragment = context.Fragments.FindDefinition(spread.Name);
+                                if (fragment != null && ShouldIncludeNode(context, fragment) && DoesFragmentConditionMatch(context, fragment.Type.Name, specificType))
+                                    CollectFields(context, specificType, fragment.SelectionSet, fields, ref visitedFragmentNames);
+                            }
+                        }
+                        else if (selection is InlineFragment inline)
+                        {
+                            // inline.Type may be null
+                            // See [2.8.2] Inline Fragments: If the TypeCondition is omitted, an inline fragment is considered to be of the same type as the enclosing context.
+                            if (ShouldIncludeNode(context, inline) && DoesFragmentConditionMatch(context, inline.Type?.Name ?? specificType.Name, specificType))
+                                CollectFields(context, specificType, inline.SelectionSet, fields, ref visitedFragmentNames);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method calculates the criterion for matching fragment definition (spread or inline) to a given graph type.
+        /// This criterion determines the need to fill the resulting selection set with fields from such a fragment.
+        /// <br/><br/>
+        /// <see href="http://spec.graphql.org/June2018/#DoesFragmentTypeApply()"/>
+        /// </summary>
+        protected virtual bool DoesFragmentConditionMatch(ExecutionContext context, string fragmentName, IGraphType type /* should be named type*/)
```

Comment:
The behavior of this method is described in the specification and should not change.

Reply:
Correct; it _should_ not change.  In fact the majority of this PR _should_ not change.  The point of this PR is to allow modifications to _any_ part of the execution strategy, regardless of whether it meets spec or not.

Reply:
Maybe we shouldn't make absolutely everything `virtual`? Some small parts can be kept private, they are fixed. I agree that it's good to make `virtual` `CollectFields` for example.